eureka2-ext/eureka2-eureka1-rest-api/src/main/java/com/netflix/eureka2/eureka1/rest/Eureka1QueryRequestHandler.java

```diff
@@ -81,57 +85,88 @@ public Eureka1QueryRequestHandler(Eureka1Configuration config, ExtensionContext
+        return registryViewCache.findApplicationsByVip(vip).flatMap(new Func1<Applications, Observable<Void>>() {
+            @Override
+            public Observable<Void> call(Applications applications) {
+                return encodeResponse(format, gzip, response, applications);
+            }
+        });
     }
 
-    private Observable<Void> secureVipGET(String secureVip, EncodingFormat format, boolean gzip, HttpServerResponse<ByteBuf> response) throws IOException {
-        Applications applications = registryViewCache.findApplicationsBySecureVip(secureVip);
-        return encodeResponse(format, gzip, response, applications);
+    private Observable<Void> secureVipGET(String secureVip, final EncodingFormat format, final boolean gzip, final HttpServerResponse<ByteBuf> response) {
+        return registryViewCache.findApplicationsBySecureVip(secureVip).flatMap(new Func1<Applications, Observable<Void>>() {
+            @Override
+            public Observable<Void> call(Applications applications) {
+                return encodeResponse(format, gzip, response, applications);
+            }
+        });
     }
 
-    private Observable<Void> instanceGetByAppAndInstanceId(String appName, String instanceId, EncodingFormat format,
-                                                           boolean gzip, HttpServerResponse<ByteBuf> response) throws IOException {
-        com.netflix.appinfo.InstanceInfo v1InstanceInfo = registryViewCache.findInstance(instanceId);
-        if (v1InstanceInfo == null) {
-            logger.info("Instance info with id {} not found", instanceId);
-            response.setStatus(HttpResponseStatus.NOT_FOUND);
-            return Observable.empty();
-        }
-        if (!appName.equalsIgnoreCase(v1InstanceInfo.getAppName())) {
-            logger.info("Instance info with id {} is associated with application {}, not {}",
-                    instanceId, v1InstanceInfo.getAppName(), appName);
-            response.setStatus(HttpResponseStatus.NOT_FOUND);
-            return Observable.empty();
-        }
-        return encodeResponse(format, gzip, response, v1InstanceInfo);
+    private Observable<Void> instanceGetByAppAndInstanceId(final String appName, final String instanceId, final EncodingFormat format,
+                                                           final boolean gzip, final HttpServerResponse<ByteBuf> response) {
+        return registryViewCache.findInstance(instanceId).materialize().toList().flatMap(
+                new Func1<List<Notification<InstanceInfo>>, Observable<Void>>() {
+                    @Override
+                    public Observable<Void> call(List<Notification<InstanceInfo>> notifications) {
+                        // If completed with error propagate it
+                        Notification<InstanceInfo> lastNotification = notifications.get(notifications.size() - 1);
```

Comment:
We will always have at least one item. We have to do this that way for proper onComplete handling.